### ISSUE TITLE
Accessing datasets as inputs in a training run

### ISSUE THREAD

**Issue by lucazav** [Azure/azureml-sdk-for-r#277]
I've just executed a training run where I had to get data from an Azure SQL DB.
I need to get the Workspace object in order to point to the specific Datastore. So I used the following code:

`ws <- load_workspace_from_config()`

My run failed with this error message:

> 
> [1] "/mnt/batch/tasks/shared/LS_root/jobs/xxxxx/yyyyyy/timeseries-clustering_1583510126_952fed96/mounts/zzzzzzz"
> Performing interactive authentication. Please follow the instructions on the terminal.
> To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code CDHTVECYU to authenticate.
> 
>  *** caught segfault ***
> address 0x8, cause 'memory not mapped'
> 
> Traceback:
>  1: py_call_impl(callable, dots$args, dots$keywords)
>  2: azureml$core$workspace$Workspace$from_config(path = path, `_file_name` = file_name)
>  3: load_workspace_from_config()
> An irrecoverable exception occurred. R is aborting now ...

Have I to use another way to get the `ws` object?

**Comment by diondrapeck** [comment]
@lucazav  - could you provide more detail? Did you open the browser link and enter the authentication code before this error happened? Or did it happen before you opened the link?

**Comment by lucazav** [comment]
@diondrapeck, I executed a training scipt in which I'm reading the object `ws` through an estimator.
So, I can't access that link. I sow the message into the _70_driver_log.txt_ file:

![image](https://user-images.githubusercontent.com/6104437/76199131-95693700-61ef-11ea-9445-08118e225679.png)

**Comment by diondrapeck** [comment]
While I'm investigating this, could you try using our `interactive_login_authentication` function? There's an example of how to pass that object into an workspace constructor in the documentation: https://azure.github.io/azureml-sdk-for-r/reference/interactive_login_authentication.html. Hopefully, that will succeed as another option for now.

**Comment by lucazav** [comment]
I tried, but I get the following error:

> could not find function "interactive_login_authentication"

**Comment by diondrapeck** [comment]
Which SDK installation are you using? That function is available in the current GitHub repo.

**Comment by diondrapeck** [comment]
Instead of re-installing to get the latest GitHub repo functions, you could also just call it directly from the Python SDK like this:

`interactive_auth <- azureml$core$authentication$InteractiveLoginAuthentication(force = force, tenant_id = tenant_id, cloud = cloud)`

**Comment by lucazav** [comment]
Yep, I already tried it:

> azureml$core$authentication$InteractiveLoginAuthentication(tenant_id = "xxxxxx")

but I get always the first kind of error:

> Performing interactive authentication. Please follow the instructions on the terminal.
> To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code CGKXX4TR7 to authenticate.
> 
>  *** caught segfault ***
> address 0x8, cause 'memory not mapped'

**Comment by diondrapeck** [comment]
To clarify, you get that error when trying `azureml$core$authentication$InteractiveLoginAuthentication(tenant_id = "xxxxxx")`? Or do you get it after that, when you try to pass it into `get_workspace`?

**Comment by lucazav** [comment]
I get it when trying the interactive authentication

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by lucazav** [comment]
Hi @diondrapeck, any news? Thanks

**Comment by diondrapeck** [comment]
Hi @lucazav, sorry for the delay. I'm trying to recreate. It looks like an issue with R rather than the Python part of the SDK; however, that error has multiple causes, so I'm trying to narrow it down to the exact one you're experienced.

In the meantime, did you try to use [`get_workspace()`](https://azure.github.io/azureml-sdk-for-r/reference/get_workspace.html) and if so, did the same error occur?

**Comment by lucazav** [comment]
The problem is that it gives that error here:

`azureml$core$authentication$InteractiveLoginAuthentication(tenant_id = 'XXX')`

and stops before executing the `get_workspace()` function.

**Comment by diondrapeck** [comment]
Could you paste a snippet of the code you're running? Just the workspace retrieval parts (personal info redacted of course).

**Comment by lucazav** [comment]
Here we are:

```
library(azuremlsdk)

sqlQuery <- "SELECT * FROM my_schema.my_table"

azureml$core$authentication$InteractiveLoginAuthentication(tenant_id = 'XXX')

ws <- get_workspace("zzz",
                      auth = auth,
                      subscription_id = "xxx",
                      resource_group = "yyy")

sql_ds <- get_datastore(ws, "my_azuresqldb")
  
dp <- data_path( sql_ds, sqlQuery )

tds <- create_tabular_dataset_from_sql_query(dp)
  
df <- load_dataset_into_data_frame(tds)
```

**Comment by diondrapeck** [comment]
Do you need the interactive login step here? What happens if you just run get_workspace without the auth param?

**Comment by lucazav** [comment]
I've just used the following code:

```
library(azuremlsdk)

sqlQuery <- "SELECT * FROM my_schema.my_table"

#azureml$core$authentication$InteractiveLoginAuthentication(tenant_id = 'XXX')

ws <- get_workspace("zzz",
                      #auth = auth,
                      subscription_id = "xxx",
                      resource_group = "yyy")

sql_ds <- get_datastore(ws, "my_azuresqldb")
  
dp <- data_path( sql_ds, sqlQuery )

tds <- create_tabular_dataset_from_sql_query(dp)
  
df <- load_dataset_into_data_frame(tds)
```
and I'm getting the following error:

> Performing interactive authentication. Please follow the instructions on the terminal.
> To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code C3Q9QJSUH to authenticate.
> 
>  *** caught segfault ***
> address 0x8, cause 'memory not mapped'
> 
> Traceback:
>  1: py_call_impl(callable, dots$args, dots$keywords)
>  2: azureml$core$Workspace$get(name, auth = NULL, subscription_id = subscription_id,     resource_group = resource_group)
>  3: doTryCatch(return(expr), name, parentenv, handler)
>  4: tryCatchOne(expr, names, parentenv, handlers[[1L]])
>  5: tryCatchList(expr, classes, parentenv, handlers)
>  6: tryCatch({    azureml$core$Workspace$get(name, auth = NULL, subscription_id = subscription_id,         resource_group = resource_group)}, error = function(e) {    if (grepl("No workspaces found with name=", e$message, )) {        NULL    }    else {        stop(message(e))    }})
>  7: get_workspace("xxx", subscription_id = "yyy",     resource_group = "zzz")
> An irrecoverable exception occurred. R is aborting now ...
> bash: line 1:   119 Segmentation fault      (core dumped) "Rscript" "training_clustering.R" "--data_folder" "$AZUREML_DATAREFERENCE_xxx_download"
> 2020/03/11 17:25:47 mpirun version string: {
> mpirun (Open MPI) 3.1.2

**Comment by lucazav** [comment]
I'm also trying to use the MSIAuthentication in this way (first from my Compute instance, not using the upon mentioned estimator):

```
auth <- azureml$core$authentication$MsiAuthentication()

ws <- get_workspace("zzz",
                      auth = auth,
                      subscription_id = "xxx",
                      resource_group = "yyy")
```

but I'm getting the following error:

> HTTPError: Error in value[[3L]](cond) : 

@mx-iao , @diondrapeck any chance to speed up the resolution of this issue? I'm stuck with my solution now. Thank you very much.

**Comment by diondrapeck** [comment]
@lucazav - I'm sorry, I can't reproduce. Could you try installing the latest GitHub repo? I have no issues with interactive auth using the GitHub install.

**Comment by lucazav** [comment]
Hi @diondrapeck, I've just updated my R environment on-prem using this code:

```
remotes::install_github('https://github.com/Azure/azureml-sdk-for-r')

azuremlsdk::install_azureml(version = "1.0.85", remove_existing_env = TRUE)
```
Then I've run the estimator in the following way:

```
library(azuremlsdk)

setwd("C:\\...\\azureml-folder")

auth <- interactive_login_authentication(tenant_id="xxx")

ws <- get_workspace("my_workspace",
                    auth = auth,
                    subscription_id = "my_subscription_id", 
                    resource_group = "my_resourcegroup")

ds <- get_datastore(ws, "my_download")

experiment_name <- "timeseries-clustering"
exp <- experiment(ws, experiment_name)

cluster_name <- "rcluster"
compute_target <- get_compute(ws, cluster_name = cluster_name)

if (is.null(compute_target)) {
  vm_size <- "STANDARD_D2_V2" 
  compute_target <- create_aml_compute(workspace = ws,
                                       cluster_name = cluster_name,
                                       vm_size = vm_size,
                                       max_nodes = 2)
  
  wait_for_provisioning_completion(compute_target, show_output = TRUE)
}

target_path <- "./"

est <- estimator(source_directory = ".",
                 entry_script = "training_test.R",
                 cran_packages = c("tidyverse", "dtwclust", "clue", "smooth"),
                 script_params = list("--data_folder" = ds$path(target_path)),
                 compute_target = compute_target )

run <- submit_experiment(exp, est)
```
The `training_test.R` file contains the following code:

```
library(azuremlsdk)

sqlQuery <- "SELECT * FROM my_schema.my_table AS A"

auth <- 
  azureml$core$authentication$InteractiveLoginAuthentication(tenant_id = 'xxx')

ws <- get_workspace("my_workspace",
                    auth = auth,
                    subscription_id = "my_subscription_id",
                    resource_group = "my_resourcegroup")

sql_ds <- get_datastore(ws, "my_azuresqldb")

dp <- data_path( sql_ds, sqlQuery )

tds <- create_tabular_dataset_from_sql_query(dp)

df <- load_dataset_into_data_frame(tds)
```

My run fails with the following error I get from the `70_driver_log.txt` file:

> Performing interactive authentication. Please follow the instructions on the terminal.
> To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code C6CMJHWEP to authenticate.
> 
>  *** caught segfault ***
> address 0x8, cause 'memory not mapped'
> 
> Traceback:
>  1: py_call_impl(callable, dots$args, dots$keywords)
>  2: azureml$core$authentication$InteractiveLoginAuthentication(tenant_id = "xxx")
> An irrecoverable exception occurred. R is aborting now ...
> bash: line 1:   118 Segmentation fault      (core dumped) "Rscript" "training_test.R" "--data_folder" "$AZUREML_DATAREFERENCE_my_download"
> 2020/03/16 22:24:46 mpirun version string: {
> mpirun (Open MPI) 3.1.2
> 
> Report bugs to http://www.open-mpi.org/community/help/
> }

Is there anything I'm doing in the wrong way?
Thanks.

**Comment by diondrapeck** [comment]
@lucazav - Thanks for posting the training script - that helps a ton! So you're trying to run interactive authentication from inside of your remote run, and I believe that's what's causing your issue. You don't need to authorize again nor retrieve a workspace once you've kicked off your run because your run is part of an experiment that you created using your workspace that you've been authorized to use. Re-writing those steps inside the script isn't necessary; the only thing that needs to be in your training script is the code you want run for training or your experiment. 

It's on our list to add a datastore/dataset vignette example, but in the meantime @mx-iao could you please hop in and give @lucazav guidance on how to best handle his datastores and sets for this run example? He'll need to re-write his script "training_test.R".

**Comment by mx-iao** [comment]
@lucazav Yes thanks for posting the training code - let me provide some guidance for you on dataset usage. I am currently in the process of writing up a datasets usage vignette, which I should be able to get in by the end of the week.

You want to pass in your dataset(s) required in the training script as part of the `inputs` parameter of the `estimator` function like so:
```
sql_ds <- get_datastore(ws, "my_azuresqldb")
dp <- data_path( sql_ds, sqlQuery )
tds <- create_tabular_dataset_from_sql_query(dp)

est <- estimator(...<other arguments>..., 
           inputs = list(dataset_consumption_config("my-train-ds", tds, mode = "direct") )
```
See [`dataset_consumption_config()`](https://azure.github.io/azureml-sdk-for-r/reference/dataset_consumption_config.html) for full set of arguments. The `name` (e.g. in above example "my-train-ds" is a friendly name for referring to your dataset that you will use within your training script. This can be different than the name that the dataset is registered under in the workspace. `mode` is either direct/mount/or download - for tabular datasets you can only do direct mode, which is the default. For file datasets the other options or mounting or download to compute target.

Then, in your training script, you can do the following:
```
library(azuremlsdk)

# retrieve the dataset you passed into `inputs` via its friendly name
tds <- get_input_dataset_from_run("my-train-ds")

# load into data frame
df <- load_dataset_into_data_frame(tds)

# rest of your training code
```

For some more information on Azure ML datasets see this article: https://docs.microsoft.com/en-us/azure/machine-learning/how-to-train-with-datasets. Code snippets there are in Python but the constructs are the same.

**Comment by lucazav** [comment]
@mx-iao , I tried what you suggested in this way:

**Estimator**
```
library(azuremlsdk)

auth <- interactive_login_authentication(tenant_id="xxx")

ws <- get_workspace("xxx",
                    auth = auth,
                    subscription_id = "yyy", 
                    resource_group = "zzz")

sql_ds <- get_datastore(ws, "my_azuresqldb")

sqlQuery <- "SELECT TOP 1000 * FROM my_schema.my_table"

dp <- data_path( sql_ds, sqlQuery )

tds <- create_tabular_dataset_from_sql_query(dp)

ds <- get_datastore(ws, "my_download")

experiment_name <- "timeseries-clustering"
exp <- experiment(ws, experiment_name)

cluster_name <- "rcluster"
compute_target <- get_compute(ws, cluster_name = cluster_name)

if (is.null(compute_target)) {
  vm_size <- "STANDARD_D2_V2" 
  compute_target <- create_aml_compute(workspace = ws,
                                       cluster_name = cluster_name,
                                       vm_size = vm_size,
                                       max_nodes = 2)
  
  wait_for_provisioning_completion(compute_target, show_output = TRUE)
}

target_path <- "./"

est <- estimator(source_directory = ".",
                 entry_script = "training_test.R",
                 cran_packages = c("tidyverse", "dtwclust", "clue", "smooth"),
                 script_params = list("--data_folder" = ds$path(target_path)),
                 inputs = list(dataset_consumption_config("my_tabular_ds", tds, mode = "direct")),
                 compute_target = compute_target )

run <- submit_experiment(exp, est)
```

**training_test.R**
```
library(azuremlsdk)

tds <- get_input_dataset_from_run("my_tabular_ds")

df <- load_dataset_into_data_frame(tds)
```

but I'm getting the following error:

> Error in dataset$to_pandas_data_frame() : attempt to apply non-function
> Calls: load_dataset_into_data_frame
> Execution halted
> 2020/03/17 11:47:50 mpirun version string: {
> mpirun (Open MPI) 3.1.2
> 
> Report bugs to http://www.open-mpi.org/community/help/
> }

**Comment by mx-iao** [comment]
@lucazav can you update your R SDK to the latest from GitHub? There was a fix for this that just got checked in. I would suggest using a GitHub install for now as we are checking bug fixes and enhancements; we are planning another CRAN release soon.

**Comment by lucazav** [comment]
@mx-iao I've just tried to update the R SDK I have on my Compute Instance, but it was already updated:

`remotes::install_github('https://github.com/Azure/azureml-sdk-for-r', force = "TRUE")`

> Skipping install of 'azuremlsdk' from a github remote, the SHA1 (3385f237) has not changed since last install.
>   Use `force = TRUE` to force installation
> 

I run my code again for a double-check and I got the upon mentioned error again.
Anything I'm missing?

**Comment by mx-iao** [comment]
Ah sorry - misread your comment. I see, your R SDK is up to date but the SDK used in the default container built for the remote run is installing the CRAN latest. Can you specify this to your `r_environment` to get the GitHub install `github_packages=[“Azure/azureml-sdk-for-r”]` . 

We are working on finalizing our cadence for CRAN release and will take this as input, please try the above workaround right now. Thank you!

**Comment by lucazav** [comment]
@mx-iao I've just used the following code:

```
est <- estimator(source_directory = ".",
                 entry_script = "training_test.R",
                 cran_packages = c("tidyverse", "dtwclust", "clue", "smooth"),
                 github_packages= c("Azure/azureml-sdk-for-r"),
                 script_params = list("--data_folder" = ds$path(target_path)),
                 inputs = list(dataset_consumption_config("my_tabular_ds", tds, mode = "direct")),
                 compute_target = compute_target )

run <- submit_experiment(exp, est)
```

but I'm getting this error into the `20_image_build_log.txt` file:

> > devtools::install_github('Azure/azureml-sdk-for-r')
> [91mError in loadNamespace(name) : there is no package called ‘devtools’
> Calls: :: ... loadNamespace -> withRestarts -> withOneRestart -> doWithOneRestart
> Execution halted
> [0mThe command '/bin/sh -c R -e "devtools::install_github('Azure/azureml-sdk-for-r')"' returned a non-zero code: 1
> 2020/03/17 18:07:51 Container failed during run: acb_step_0. No retries remaining.
> failed to run step ID: acb_step_0: exit status 1
> 
> Run ID: cba failed after 21m53s. Error: failed during run, err: exit status 1

**Comment by mx-iao** [comment]
Ah you will also need `devtools` package too in order to install; can you specify that to the `cran_packages` too? 

@heemanshusuri we should include devtools as a default package too for these types of scenarios

**Comment by lucazav** [comment]
@mx-iao it seems the `devtools` package can't find its dependencies:

> [0m[91mERROR: dependencies ‘usethis’, ‘git2r’ are not available for package ‘devtools’
> [0m[91m* removing ‘/opt/miniconda/lib/R/library/devtools’
> [0m[91m
> The downloaded source packages are in
> 	‘/tmp/RtmpukUGRS/downloaded_packages’
> [0m[91mUpdating HTML index of packages in '.Library'
> [0m[91mMaking 'packages.html' ...[0m[91m done
> Warning messages:
> [0m[91m1: In install.packages("devtools", repos = "https://cloud.r-project.org/") :[0m[91m
>   installation of package ‘git2r’ had non-zero exit status
> 2: In install.packages("devtools", repos = "https://cloud.r-project.org/") :[0m[91m
>   installation of package ‘gh’ had non-zero exit status
> [0m[91m3: In install.packages("devtools", repos = "https://cloud.r-project.org/") :[0m[91m
>   installation of package ‘usethis’ had non-zero exit status
> 4: In install.packages("devtools", repos = "https://cloud.r-project.org/") :[0m> 
> > 
> [91m
>   installation of package ‘devtools’ had non-zero exit status
> [0mRemoving intermediate container b571395b0c1f

I'm trying adding also the `usethis` and `git2r` packages to CRAN's ones. Is it the right way to go?

**Comment by mx-iao** [comment]
Yeah, can you try including those too? Those packages should be automatically getting installed since they're specified in Imports.. @heemanshusuri can you please take a look?

**Comment by lucazav** [comment]
Another error. Also the `gh` package was missing.
But have you ever tested this case? It seems I'm the first one who is testing the `github_packages` feature...

**Comment by mx-iao** [comment]
Yes we have - @heemanshusuri will take a deeper look and see what the underlying issue is and if we can add any more robust testing. Thanks for your patience; we are a small team and stretched quite thin at the minute.

**Comment by lucazav** [comment]
I understand. I was asking just to suggest you to reinforce the testing phase.
In the meantime I got this error when installing the `gh` package:

> [0m[91m* installing *source* package ‘gh’ ...
> [0m[91m** package ‘gh’ successfully unpacked and MD5 sums checked
> [0m[91m** using staged installation
> [0m[91m** R
> [0m[91m** byte-compile and prepare package for lazy loading
> [0mError: object ‘cli_alert_info’ is not exported by 'namespace:cli'
> Execution halted
> [91mERROR: lazy loading failed for package ‘gh’
> [0m[91m* removing ‘/opt/miniconda/lib/R/library/gh’

When installing the package git2r I get this error:

> [0m[91m* installing *source* package ‘git2r’ ...
> [0m[91m** package ‘git2r’ successfully unpacked and MD5 sums checked
> [0m[91m** using staged installation
> [0mchecking build system type... x86_64-unknown-linux-gnu
> checking host system type... x86_64-unknown-linux-gnu
> checking for gcc... gcc
> checking whether the C compiler works... yes
> checking for C compiler default output file name... a.out
> checking for suffix of executables... 
> checking whether we are cross compiling... no
> checking for suffix of object files... o
> checking whether we are using the GNU C compiler... yes
> checking whether gcc accepts -g... yes
> checking for gcc option to accept ISO C89... none needed
> checking for pkg-config... no
> configure: Package dependency requirement 'libgit2 >= 0.26.0' could not be satisfied.
> 
>   -----------------------------------------------------------------------
> 
>    Unable to find the libgit2 library on this system. Building 'git2r'
>    using the bundled source of the libgit2 library.
> 
>    To build git2r with a system installation of libgit2, please install:
>      libgit2-dev   (package on e.g. Debian and Ubuntu)
>      libgit2-devel (package on e.g. Fedora, CentOS and RHEL)
>      libgit2       (Homebrew package on OS X)
>    and try again.
> 
>    If the libgit2 library is installed on your system but the git2r
>    configuration is unable to find it, you can specify the include and
>    lib path to libgit2 with:
> 
>    given you downloaded a tar-gz archive:
>    R CMD INSTALL git2r-.tar.gz --configure-vars='INCLUDE_DIR=/path/to/include LIB_DIR=/path/to/lib'
> 
>    or cloned the GitHub git2r repository into a directory:
>    R CMD INSTALL git2r/ --configure-vars='INCLUDE_DIR=/path/to/include LIB_DIR=/path/to/lib'
> 
>    or download and install git2r in R using
>    install.packages('git2r', type='source', configure.vars='LIB_DIR=-L/path/to/libs INCLUDE_DIR=-I/path/to/headers')
> 
>    On macOS, another possibility is to let the configuration
>    automatically download the libgit2 library from the Homebrew
>    package manager with:
> 
>    R CMD INSTALL git2r-.tar.gz --configure-vars='autobrew=yes'
>    or
>    R CMD INSTALL git2r/ --configure-vars='autobrew=yes'
>    or
>    install.packages('git2r', type='source', configure.vars='autobrew=yes')
> 
>   -----------------------------------------------------------------------
> 
> 
> configure: Attempting configuration of bundled libgit2
> checking size of void*... 8
> checking for library containing inflate... [91mconfigure: error: in `/tmp/RtmpJGBW1h/R.INSTALL1f6ed8bc6f/git2r':
> [0mno
> [91mconfigure: error: 
>   ---------------------------------------------
>    The zlib library that is required to build
>    git2r was not found.
> 
>    Please install:
>      zlib1g-dev (package on e.g. Debian and Ubuntu)
>      zlib-devel (package on e.g. Fedora, CentOS and RHEL)
>    and try again.
> 
>    If the zlib library is installed on your
>    system but the git2r configuration is
>    unable to find it, you can specify the
>    include and lib path to zlib with:
>    R CMD INSTALL git2r --configure-vars='LIBS=-L/path/to/libs CPPFLAGS=-I/path/to/headers'
>   ---------------------------------------------
> See `config.log' for more details
> [0m[91mERROR: configuration failed for package ‘git2r’
> [0m[91m* removing ‘/opt/miniconda/lib/R/library/git2r’
> [0m[91m
> The downloaded source packages are in
> 	‘/tmp/RtmpeCekLx/downloaded_packages’
> [0m[91mUpdating HTML index of packages in '.Library'
> [0m[91mMaking 'packages.html' ...[0m[91m done
> Warning message:
> In install.packages("git2r", repos = "https://cloud.r-project.org/") :[0m> 
> > 
> [91m
>   installation of package ‘git2r’ had non-zero exit status
> [0mRemoving intermediate container 80949921b246

**Comment by mx-iao** [comment]
Hmm I see -  I think this issue is in scenarios where system dependencies need to be installed; we are currently determining the best way to support this w.r.t to our underlying service. @sdonohoo I believe you have worked through some similar scenarios recently - were there any quick workarounds you were able to find?

**Comment by lucazav** [comment]
Hi guys, any suggestion to not get stuck?

Also providing a custom Docker image in the meantime you'll fix the dependencies issue could be really useful for who isn't a Docker user.

Thank you very much.

**Comment by lucazav** [comment]
I create a custom Docker image as a quick workaround. You can find details [here](https://github.com/Azure/azureml-sdk-for-r/issues/260#issuecomment-601721228).

**Comment by mx-iao** [comment]
Hi @lucazav, we found a fix for this issue - our code was using `remotes` instead of `devtools` for the GitHub package installs. @heemanshusuri just checked in a fix for this yesterday #290 so if you do a new install from GitHub of the SDK you should have this fix now.
